bootloader: implement support for resolving executable symlinks on Windows

rokm opened the pull request:
Implement support for resolving true location of the executable when launched via a symbolic link on Windows. Closes #8300.

Also clean up the surrounding code (i.e., remove `pyi_path_fullpath_keep_basename` in the POSIX codepath).

rokm committed:
bootloader: remove pyi_path_fullpath_keep_basename

The `pyi_path_fullpath_keep_basename` function was introduced as part
of #3829, which sought to preserve process name of the child process in
onefile applications on POSIX OSes (other than macOS) by disabling the
full symlink resolution when resolving the executable's name.

Subsequently, #5232 re-introduced the executable name resolution via
`readlink()` on `/proc` entry, and implemented process name preservation
via `prctl` call with `PR_GET_NAME` and `PR_SET_NAME`.

At that point, the use of `pyi_path_fullpath_keep_basename˙ became
limited to the fall-back codepath where `/proc` entry resolution failed.

Later on, #7556 explicitly forced the use of this fall-back codepath
when executable is launched via `ld.so` dynamic loader, as resolution
via `/proc` entry ends up pointing to the `ld.so`. In order to handle
cases when the executable is a symbolic link, an explicit resolution
of symbolic link was added at the end of executable path resolution.

At this point, the `pyi_path_fullpath_keep_basename` became completely
redundant; even if it goes out of its way to resolve only symlinks
in the parent path components and leave the file itself alone, if
the file itself is a symbolic link, it gets resolved during the
afore-mentioned final symlink check.

So remove `pyi_path_fullpath_keep_basename` and replace its uses
with calls to `pyi_path_fullpath`, which both canonicalizes the path
and fully resolves the symlinks.

At the same time, fix error messages in `pyi_path_executable` when
`pyi_path_fullpath` fails; several of those seem to be using the
output `execfile` buffer instead of the input buffer that contains
path that is supposed to be resolved.

rokm committed:
bootloader: add symlink resolution to common part of pyi_path_executable

Move the explicit symlink resolution step from the end of POSIX
codepath in `pyi_path_executable` to the common codepath at the
end of the function. This effectively adds the explicit symlink
resolution step on Windows, although it is not fully functional
yet due to `pyi_path_fullpath` lacking symlink resolution on
Windows.

rokm committed:
bootloader: rename pyi_path_fullpath to pyi_path_resolve

Rename `pyi_path_fullpath` to `pyi_path_resolve`. Rename its input
and output path arguments to `path` and `resolved_path`, which
matches the name used on underlying POSIX `realpath()` function.
Remove the buffer size parameter, since implementation already
partly assumes that PATH_MAX characters are allowed.

rokm committed:
bootloader: resolve symlinks in Windows codepath of pyi_path_resolve

Have the Windows codepath of `pyi_path_resolve` (former
`pyi_path_fullpath`) resolve symbolic links, by replacing use of
`_wfullpath` with `GetFinalPathNameByHandleW`.

rokm committed:
bootloader: remove extended path indicator from resolved symbolic link

Have `pyi_path_resolve` remove the extended path indicator (i.e.,
the `\\?\` prefix) from resolved symbolic links on Windows. While
win32 API should handle these without problems, the prefix also
appears in `sys.executable` and onedir `sys._MEIPASS`, and that
might cause issues. Not to mention that bootloader's own homegrown
path-handling functions have very limited awareness of extended
paths.

bwoodsend commented:
Is a test out of the question? CI on Windows is already admin so we should have permission there. We can `pytest.skip()` it if the mklink raises a permission error locally.

rokm committed:
tests: add a test for bootloader fully resolving symlinked executable

Add a test for bootloader's ability to fully resolve true executable's
location when launched via a symbolic link, on all OSes.
